File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/KinOkHttpClientFactory.java
package kin.sdk;

import java.io.IOException;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class KinOkHttpClientFactory {

    private static final String KIN_SDK_ANDROID_VERSION_HEADER = "kin-sdk-android-version";
    private static final int TRANSACTIONS_TIMEOUT = 30;

    private String versionName;

    public final OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(TRANSACTIONS_TIMEOUT, TimeUnit.SECONDS)
            .writeTimeout(TRANSACTIONS_TIMEOUT, TimeUnit.SECONDS)
            .readTimeout(TRANSACTIONS_TIMEOUT, TimeUnit.SECONDS)
            .addInterceptor(new KinHeaderInterceptor())
            .addInterceptor(new RetryInterceptor())
            .build();

    public final OkHttpClient testClient = new OkHttpClient.Builder()
            .connectTimeout(TRANSACTIONS_TIMEOUT, TimeUnit.SECONDS)
            .writeTimeout(TRANSACTIONS_TIMEOUT, TimeUnit.SECONDS)
            .readTimeout(TRANSACTIONS_TIMEOUT, TimeUnit.SECONDS)
            .addInterceptor(new KinHeaderInterceptor())
            .build();

    public KinOkHttpClientFactory(String versionName) {
        this.versionName = versionName;
    }

    private class KinHeaderInterceptor implements Interceptor {

        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            request = request.newBuilder()
                    .addHeader(KIN_SDK_ANDROID_VERSION_HEADER, versionName)
                    .build();
            return chain.proceed(request);
        }
    }

    private class RetryInterceptor implements Interceptor {

        private RetryStrategy retryStrategy = new ExponentialBackoffRetryStrategy();

        @Override
        public Response intercept(Chain chain) throws IOException {
            Response response = chain.proceed(chain.request());

            int attempt = 0;
            while (retryStrategy.shouldRetry(response, attempt++)) {
                response.close();

                Request newRequest = response.request()
                        .newBuilder()
                        .build();

                response = retryStrategy.retry(newRequest, attempt, chain::proceed);
            }

            return response;
        }
    }

    public interface RetryStrategy {

        boolean shouldRetry(Response response, int attempt);

        Response retry(Request newRequest, int attempt, RequestFunction requestFunction) throws IOException;
    }

    public interface RequestFunction {

        Response run(Request request) throws IOException;
    }

    public class ExponentialBackoffRetryStrategy implements RetryStrategy {

        private static final long BACKOFF_DELAY_MILLIS = 1000;
        private static final long BACK0FF_FACTOR = 2;
        private static final long RETRY_MAX_ATTEMPTS = 4;

        private final long delayMillis;
        private final long backoffFactor;
        private final long maxAttempts;

        ExponentialBackoffRetryStrategy() {
            this(BACKOFF_DELAY_MILLIS, BACK0FF_FACTOR, RETRY_MAX_ATTEMPTS);
        }

        ExponentialBackoffRetryStrategy(long delayMillis, long backoffFactor, long maxAttempts) {
            this.delayMillis = delayMillis;
            this.backoffFactor = backoffFactor;
            this.maxAttempts = maxAttempts;
        }

        public boolean shouldRetry(Response response, int attempt) {
            return response.code() != 200 && attempt < maxAttempts;
        }

        public Response retry(Request newRequest, int attempt, RequestFunction requestFunction) throws IOException {
            try {
                Thread.sleep(delayMillis * (attempt != 0 ? backoffFactor : 1) * attempt);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            Logger.getLogger("").info("Retry Attempt " + attempt);
            return requestFunction.run(newRequest);
        }
    }

    public class NoRetryStrategy implements RetryStrategy {

        @Override
        public boolean shouldRetry(okhttp3.Response response, int attempt) {
            return false;
        }

        @Override
        public Response retry(Request newRequest, int attempt, RequestFunction requestFunction) throws IOException {
            throw new RuntimeException("No retry strategy set");
        }
    }
}

File: kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/internal/KinOkHttpClientFactory.kt
package kin.sdk.internal

import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit
import java.util.logging.Logger

class KinOkHttpClientFactory(versionName: String) {

    private companion object {
        const val TRANSACTIONS_TIMEOUT = 30
    }

    val client = OkHttpClient.Builder()
        .connectTimeout(TRANSACTIONS_TIMEOUT.toLong(), TimeUnit.SECONDS)
        .writeTimeout(TRANSACTIONS_TIMEOUT.toLong(), TimeUnit.SECONDS)
        .readTimeout(TRANSACTIONS_TIMEOUT.toLong(), TimeUnit.SECONDS)
        .addInterceptor(KinHeaderInterceptor(versionName))
        .addInterceptor(RetryInterceptor())
        .build()

    val testClient = OkHttpClient.Builder()
        .connectTimeout(TRANSACTIONS_TIMEOUT.toLong(), TimeUnit.SECONDS)
        .writeTimeout(TRANSACTIONS_TIMEOUT.toLong(), TimeUnit.SECONDS)
        .readTimeout(TRANSACTIONS_TIMEOUT.toLong(), TimeUnit.SECONDS)
        .addInterceptor(KinHeaderInterceptor(versionName))
        .build()
}

private class KinHeaderInterceptor(private val versionName: String) : Interceptor {

    private companion object {
        const val KIN_SDK_ANDROID_VERSION_HEADER = "kin-sdk-android-version"
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        var request = chain.request()
        request = request.newBuilder()
            .addHeader(KIN_SDK_ANDROID_VERSION_HEADER, versionName)
            .build()
        return chain.proceed(request)
    }
}

private class RetryInterceptor : Interceptor {

    private val retryStrategy = ExponentialBackoffRetryStrategy()

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        var response = chain.proceed(chain.request())

        var attempt = 0
        while (retryStrategy.shouldRetry(response, attempt++)) {
            response.close()

            val newRequest = response.request()
                .newBuilder()
                .build()

            response = retryStrategy.retry(newRequest, attempt, object : RequestFunction {
                override fun run(request: Request): Response {
                    return chain.proceed(request)
                }
            })
        }

        return response
    }
}

private interface RetryStrategy {

    fun shouldRetry(response: Response, attempt: Int): Boolean

    @Throws(IOException::class)
    fun retry(newRequest: Request, attempt: Int, requestFunction: RequestFunction): Response
}

private interface RequestFunction {

    @Throws(IOException::class)
    fun run(request: Request): Response
}

private class ExponentialBackoffRetryStrategy @JvmOverloads internal constructor(private val delayMillis: Long = BACKOFF_DELAY_MILLIS, private val backoffFactor: Long = BACK0FF_FACTOR, private val maxAttempts: Long = RETRY_MAX_ATTEMPTS) : RetryStrategy {

    private companion object {
        const val BACKOFF_DELAY_MILLIS: Long = 1000
        const val BACK0FF_FACTOR: Long = 2
        const val RETRY_MAX_ATTEMPTS: Long = 4
    }

    override fun shouldRetry(response: Response, attempt: Int): Boolean {
        return response.code() != 200 && attempt < maxAttempts
    }

    @Throws(IOException::class)
    override fun retry(newRequest: Request, attempt: Int, requestFunction: RequestFunction): Response {
        try {
            Thread.sleep(delayMillis * (if (attempt != 0) backoffFactor else 1) * attempt.toLong())
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }

        Logger.getLogger("").info("Retry Attempt $attempt")
        return requestFunction.run(newRequest)
    }
}
